parmed/modeller/residue.py

```diff
@@ -329,6 +377,100 @@ def fix_charges(self, to=None, precision=4):
 
         return self
 
+    def apply_patch(self, patch, precision=4):
+        """
+        Apply the specified PatchTemplate to the ResidueTemplate.
+
+        This only handles patches that affect a single residue.
+
+        An exception is thrown if patch is incompatible because
+        * The patch specifies that an atom is to be deleted that doesn't exist in the residue
+        * A bond specified as being added in the patch does not have both atom names present after adding/deleting atoms from the patch
+        * The new net charge is not integral to the specified precision
+
+        Parameters
+        ----------
+
+        patch : PatchTemplate
+            The patch to apply to this residue
+
+        precision : int, optional
+            Each valid patch should be produce a net charge that is integral to
+            this many decimal places.
+            Default is 4
+
+        Returns
+        -------
+
+        residue : ResidueTemplate
+            A new ResidueTemplate corresponding to the patched residue is returned.
+            The original remains unmodified.
+
+        """
+        # Create a copy
+        # TODO: Once ResidueTemplate.from_residue() actually copies all info, use that instead?
+        residue = _copy.copy(self)
+        # Delete atoms
+        for atom_name in patch.delete_atoms:
+            residue.delete_atom(atom_name)
+        # Add or replace atoms
+        for atom in patch.atoms:
+            if atom.name in residue:
+                # Overwrite type and charge
+                residue[atom.name].type = atom.type
+                residue[atom.name].charge = atom.charge
+            else:
+                residue.add_atom(Atom(name=atom.name, type=atom.type, charge=atom.charge))
+        # Add bonds
+        for (atom1_name, atom2_name, order) in patch.add_bonds:
```

Comment:
I though "add_bonds" refers to verb rather a noun?

Reply:
Good point -- maybe change this to `added_bonds`?

Although in this case this patch indicates a verb should take place -- bonds in this list need to be added when the patch is activated.

Reply:
> Although in this case this patch indicates a verb should take place -- bonds in this list need to be added when the patch is activated.

Right. That's why I went with `delete_atoms` and `delete_impropers`. I could change these too, but the change should probably be consistent.

Reply:
In the line 423, we are using `residue.add_atom(...)`
but in the line 425, we do `... in patch.add_bonds`. It's very confusing.

Reply:
> In the line 423, we are using residue.add_atom(...)
but in the line 425, we do ... in patch.add_bonds. It's very confusing.

One is a method, another is a field. `PatchTemplate` has three fields that instruct the patch on what needs to be done:
* `add_atoms`
* `delete_bonds`
* `delete_impropers`

These are three *actions* that the patch must perform. It performs one of these actions using the `ResidueTemplate.add_atom()` method to add an atom. But to add bonds by name, we must use a different scheme because parmed's `ResidueTemplate.bonds` list cannot contain a list of atom names---they must be `Atom` objects. Patches are allowed to specify names of atoms that are not defiend in the patch as part of bonds being added. This is the clearest workaround that has the smallest impact on the rest of the code.